com.unity.sg2/Editor/GraphDeltaRegistry/Generation/Interpreter.cs

```diff
@@ -20,6 +21,573 @@ namespace UnityEditor.ShaderGraph.Generation
+            {
+                public PropertyBlockUsage propertyBlockUsage;
+                public DataSource dataSource;
+                public string displayName;
+                public string defaultValue;
+                public StructField field;
+            }
+
+            private class VariableData
+            {
+                public string desiredVariableName;
+                public string actualVariableName;
+                public ShaderType shaderType;
+                public string initializationString;
+                public bool inlineInitialization;
+            }
+
+            private struct VariableReference
+            {
+                private int m_hash;
+                public VariableReference(VariableData variable)
+                {
+                    m_hash = HashCode.Combine(variable.desiredVariableName.GetHashCode(StringComparison.Ordinal),
+                                              variable.actualVariableName?.GetHashCode(StringComparison.Ordinal),
+                                              variable.shaderType,
+                                              variable.initializationString?.GetHashCode(StringComparison.Ordinal));
+                }
+            }
+
+            private struct FunctionReference
+            {
+                private int m_hash;
+                public FunctionReference(ShaderFunction function)
+                {
+                    int tmp = 0;
+                    foreach(var param in function.Parameters)
+                    {
+                        tmp += param.Name.GetHashCode(StringComparison.Ordinal) ^ param.Type.Name.GetHashCode(StringComparison.Ordinal);
+                    }
+                    m_hash = HashCode.Combine(function.Name.GetHashCode(StringComparison.Ordinal),
+                                              function.Body.GetHashCode(StringComparison.Ordinal),
+                                              tmp);
+                }
+            }
+
+            public struct ProcessingResult
+            {
+                public IEnumerable<ShaderType> usedTypes;
+                public IEnumerable<ShaderFoundry.IncludeDescriptor> includes;
+                public IEnumerable<ShaderFunction> functions;
+                public ShaderFunction entryPointFunction;
+            }
+
+            private Dictionary<VariableReference, VariableData> variableReferences;
+            private Dictionary<VariableReference, int> variableCollisions;
+
+            private Dictionary<ElementID, VariableReference> portToVariableLookup;
+            private HashSet<VariableReference> initialized;
+
+            private Dictionary<FunctionReference, ShaderFunction> functionReferences;
+            private HashSet<ShaderFoundry.IncludeDescriptor> includeReferences;
+            private HashSet<ShaderType> typeReferences;
+
+            private ShaderFunction.Builder mainFunction;
+            private HashSet<VariableReference> inputs;
```

Comment:
`inputs` isn't used anywhere other than in `WriteFunctionCall` right now, could it possibly be made local? 

Similar comments would apply for the other variables here, lot of state being retained (for good reason of course, lots to track) but making whatever possible a local variable as opposed to a member variable would be good

Reply:
currently only used in WriteFunctionCall, but it might be needed in variable registering since we conditionally promote variables. if that turns out not to be true I will change